src/julia_internal.h

```diff
@@ -593,6 +600,25 @@ STATIC_INLINE size_t jl_vararg_length(jl_value_t *v) JL_NOTSAFEPOINT
     return jl_unbox_long(len);
 }
 
+// check whether the specified number of arguments is compatible with the
+// specified number of paramters of the tuple type
+STATIC_INLINE int jl_tupletype_length_compat(jl_value_t *v, size_t nargs) JL_NOTSAFEPOINT
```

Comment:
I think this can go near the callsite (src/opaque_closure.c), since we don't want to encourage too much duplication of the subtyping code functionality. Indeed, I think eventually (future work) we should have this type-assert / method-error use the existing subtyping (specifically, `jl_tuple_isa`).